File: 0501-1000/0901_StockSpanner.py
class StockSpanner:
    def __init__(self):
        self.stack = []

    def next(self, price: int) -> int:
        if not self.stack:
            self.stack.append((price, 1))
        elif price < self.stack[-1][0]:
            self.stack.append((price, 1))
        else:
            span = 1
            while self.stack and price >= self.stack[-1][0]:
                span += self.stack[-1][1]
                self.stack.pop()
            self.stack.append((price, span))
        return self.stack[-1][1]

# Your StockSpanner object will be instantiated and called as such:
# obj = StockSpanner()
# param_1 = obj.next(price)

# Alternative solution

class StockSpanner:
    def __init__(self):
        self.stack = []

    def next(self, price: int) -> int:
        span = 1

        while self.stack and self.stack[-1][0] <= price:
            pr, sp = self.stack.pop()
            span += sp
        
        self.stack.append((price, span))

        return span

File: 0501-1000/0901_StockSpanner.java
class StockSpanner {
    private Stack<List<Integer>> stack;

    public StockSpanner() {
        this.stack = new Stack<>();
    }
    
    public int next(int price) {
        int count = 1;

        while (!stack.empty() && stack.peek().get(0)<=price) {
            count += stack.pop().get(1);
        }

        stack.push(List.of(price, count));
        return count;
    }
}

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner obj = new StockSpanner();
 * int param_1 = obj.next(price);
 */
